third_party/beam/sdks/go/pkg/beam/runners/direct/gbk.go

```diff
@@ -57,88 +108,248 @@ func (n *CoGBK) StartBundle(ctx context.Context, id string, data exec.DataManage
 }
 
 func (n *CoGBK) ProcessElement(ctx context.Context, elm exec.FullValue, _ ...exec.ReStream) error {
-	index := elm.Elm.(int)
-	value := elm.Elm2.(exec.FullValue)
+	index := elm.Elm.(int)             // index of Inputs
+	value := elm.Elm2.(exec.FullValue) // actual KV<K,V>
 
-	for _, w := range elm.Windows {
-		ws := []typex.Window{w}
+	fullKey := exec.FullValue{Elm: value.Elm, Timestamp: value.Timestamp}  // strip V from KV<K,V>
+	fullVal := exec.FullValue{Elm: value.Elm2, Timestamp: value.Timestamp} // strip K from KV<K,V>
 
-		var buf bytes.Buffer
-		if err := n.enc.Encode(exec.FullValue{Elm: value.Elm}, &buf); err != nil {
-			return fmt.Errorf("failed to encode key %v for CoGBK: %v", elm, err)
-		}
-		if err := n.wEnc.Encode(ws, &buf); err != nil {
-			return fmt.Errorf("failed to encode window %v for CoGBK: %v", w, err)
-		}
-		key := buf.String()
+	var buf bytes.Buffer
+	if err := n.keyEnc.Encode(fullKey, &buf); err != nil {
+		return fmt.Errorf("failed to encode key %v for CoGBK: %v", elm, err)
+	}
+	key := append([]byte{}, buf.Bytes()...)
+	buf.Reset()
+	if err := n.valEnc[index].Encode(fullVal, &buf); err != nil {
+		return fmt.Errorf("failed to encode val %v for CoGBK: %v", elm, err)
+	}
+	val := buf.Bytes()
 
-		g, ok := n.m[key]
-		if !ok {
-			g = &group{
-				key:    exec.FullValue{Elm: value.Elm, Timestamp: value.Timestamp, Windows: ws},
-				values: make([][]exec.FullValue, len(n.Edge.Input)),
-			}
-			n.m[key] = g
-		}
-		g.values[index] = append(g.values[index], exec.FullValue{Elm: value.Elm2, Timestamp: value.Timestamp})
+	shard := int(crc32.ChecksumIEEE(key)) % len(n.sorters)
+	n.mu[shard].Lock()
+	defer n.mu[shard].Unlock()
+	return n.sorters[shard].Add(&iterValue{
+		Key:   key,
+		Index: index,
+		Value: val,
+	})
+}
+
+type iterStream struct {
+	n *CoGBK
+
+	iter disksort.Iterator
+
+	key []byte
+	idx int
+
+	buffer *iterValue // first element of this/next shard
+	prev   *iterStream
+
+	opened bool
+}
+
+func (i *iterStream) Open() exec.Stream {
+	// TODO(schroederc): add disksort support for reiteration
+	if i.opened {
+		panic("disksort CoGBK does not support ReStreams")
 	}
-	return nil
+	i.opened = true
+	return i
 }
 
-func (n *CoGBK) FinishBundle(ctx context.Context) error {
-	for key, g := range n.m {
-		values := make([]exec.ReStream, len(g.values))
-		for i, list := range g.values {
-			values[i] = &exec.FixedReStream{Buf: list}
+func (i *iterStream) Read() (exec.FullValue, error) {
+	iv, err := i.next()
+	if err != nil {
+		return exec.FullValue{}, err
+	}
+	val, err := i.n.valDec[i.idx].Decode(bytes.NewBuffer(iv.Value))
+	if err != nil {
+		return exec.FullValue{}, fmt.Errorf("error decoding value: %v", err)
+	}
+	return val, nil
+}
+
+func (i *iterStream) next() (*iterValue, error) {
+	if i.prev != nil {
+		var prevItems int
```

Comment:
unused prevItems?

I mean I see you incrementing it, but then... nothing

Reply:
✓